Multi-target Json and Logger Source Generators between Roslyn v3.9 and v4.0

eerhardt opened the pull request:
Add the ability to target both Roslyn v3.9 and v4.0 in our Logging source generator. This is only enabled in the NuGet packages. For the ref-packs (both NetCoreApp and AspNetCore), we will only ship the `4.0` targeting versions of the generators. In order to target `net6.0`, customers must use VS 2022 which has Roslyn 4.0.

This is just an early draft to get initial feedback on the approach. The structure created here follows the proposal in https://github.com/dotnet/sdk/issues/20355.

If the SDK proposal isn't accepted for 6.0, my plan is to ship `.targets` files in both NuGet packages which will select the correct source generator version using custom logic. In both cases, we will need these changes to build the source generator twice - once for each Roslyn version we want to target.

Once we settle on an approach here, I can make the same changes to the System.Text.Json source generator.

eerhardt committed:
Multi-target LoggerMessageGenerator between Roslyn v3.9 and v4.0

ViktorHofer reviewed (approved):
The msbuild and the project changes as well as the transport package change LGTM.

eerhardt commented:
@ViktorHofer - one place I was wondering about was the changes to the `.sln` file. I know we generate our .sln files. Will the sln generator handle these changes?

eerhardt commented on src/libraries/Directory.Build.targets:
```diff
@@ -260,6 +260,7 @@
           Returns="@(_AnalyzerPackFile)">
     <PropertyGroup>
       <_analyzerPath>analyzers/dotnet</_analyzerPath>
+      <_analyzerPath Condition="'$(AnalyzerRoslynVersion)' != ''">$(_analyzerPath)/roslyn$(AnalyzerRoslynVersion)</_analyzerPath>
```
Right now: nothing. But check out https://github.com/dotnet/sdk/issues/20355 for more information.

ViktorHofer commented:
> @ViktorHofer - one place I was wondering about was the changes to the .sln file. I know we generate our .sln files. Will the sln generator handle these changes?

Feel free to check the solution file format change in as it probably will be a some time until we regenerate them again. Regarding your question, yes slngen discovers these references as [`AnalyzerReference` items are just slightly modified ProjectReferences](https://github.com/dotnet/runtime/blob/933988c35c172068652162adf6f20477231f815e/src/libraries/Directory.Build.targets#L244-L256). Because of that, the analyzers are part of the depedency graph and slngen sees them.

eerhardt committed:
Include a .targets file in NuGet packages which will select the correct analyzer assembly depending on which Roslyn version will be used to compile.

ericstj commented on eng/MultiTargetRoslynComponent.targets.template:
```diff
@@ -0,0 +1,41 @@
+          DependsOnTargets="_{TargetPrefix}ResolveRoslynApiMajorVersion">
+    <ItemGroup>
+      <!-- clear _CurrentAnalyzer, in case some other target is using it. -->
+      <_CurrentAnalyzer Remove="@(_CurrentAnalyzer)" />
```
Why didn't you just use `{TargetPrefix}` to give this item a namespace?

eerhardt replied:
Fixed.

ericstj commented on eng/MultiTargetRoslynComponent.targets.template:
```diff
@@ -0,0 +1,41 @@
+
+  <Target Name="_{TargetPrefix}RoslynComponentMultiTargeting" 
+          Condition="'$(SupportsRoslynComponentVersioning)' != 'true'" 
+          AfterTargets="ResolveAssemblyReferences" 
```
This might be too late.  Wouldn't conflict resolution run before this and drop one of them?  I guess that would depend on if we gave them the same name or not.  Actually maybe we need to give different names per @sharwell's sxs versioning suggestion.

eerhardt replied:
In my testing, it isn't too late. `_HandlePackageFileConflicts` outputs:

`Encountered conflict between 'Analyzer:C:\Users\eerhardt\.nuget\packages\system.text.json\7.0.0-dev\analyzers\dotnet\roslyn3.9\cs\System.Text.Json.SourceGeneration.dll' and 'Analyzer:C:\Users\eerhardt\.nuget\packages\system.text.json\7.0.0-dev\analyzers\dotnet\roslyn4.0\cs\System.Text.Json.SourceGeneration.dll'. Could not determine winner due to equal file and assembly versions.`

So both Analyzers are still around during `ResolveAssemblyReferences`.

However, I can move this up to be `AfterTargets="ResolvePackageDependenciesForBuild"`. Or potentially even `ResolveLockFileAnalyzers`. Thoughts?

ericstj replied:
We need your logic to run before conflict resolution.  The fact that you are seeing `Encountered conflict ` means it's not early enough.  Your logic needs to behave "as if" it were NuGet making the decision.  `AfterTargets` the target(s) that raise(s) these analyzers from NuGet packages would be best if it is / they are public.

eerhardt replied:
Both `ResolvePackageDependenciesForBuild` and `ResolveLockFileAnalyzers` happen before conflict resolution runs. And `ResolvePackageDependenciesForBuild` depends on `ResolveLockFileAnalyzers`.

The flow, as I understand it is:

1. [ResolvePackageAssets](https://github.com/dotnet/sdk/blob/17181d34b143cbf4b957abdaabb1d2fb116dc612/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L246-L275) outputs the analyzers in the project.assets.json to the `@(ResolvedAnalyzers)` item.
2. [ResolveLockFileAnalyzers](https://github.com/dotnet/sdk/blob/17181d34b143cbf4b957abdaabb1d2fb116dc612/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L485-L490) depends on `ResolvePackageAssets` and turns `@(ResolvedAnalyzers)` into `@(Analyzer)`
3. [ResolvePackageDependenciesForBuild](https://github.com/dotnet/sdk/blob/17181d34b143cbf4b957abdaabb1d2fb116dc612/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L135-L149) depends on `ResolveLockFileAnalyzers`, and doesn't have the "LockFile" in the name.

Another thing occurred to me as I wrote this - none of these targets occur with a non-SDK .csproj that uses `PackageReference`. Do you think that is a scenario we need to support?

ericstj replied:
> none of these targets occur with a non-SDK .csproj that uses PackageReference. Do you think that is a scenario we need to support?

I believe that scenario would still raise analyzers and run in VS 2022, so yes.  Our packages support install in classic .NETFramework projects.  The nice thing about `AfterTargets` is that it's a noop if the target doesn't exist, and it'll only run once for a given project/global-property-set.

eerhardt replied:
Thanks for the help. I changed this target to run `AfterTargets="ResolvePackageDependenciesForBuild;ResolveNuGetPackageAssets"`, which should address these concerns.

eerhardt committed:
Multi-target JsonSourceGenerator between Roslyn v3.9 and v4.0

eerhardt committed:
Fix restore

eerhardt committed:
Update NuGet package MSBuild logic to detect when SupportsRoslynComponentVersioning is not available, and use the lowest analyzer available.

eerhardt commented:
I believe this is ready for review, assuming https://github.com/dotnet/sdk/pull/20793 gets approved and merged for 6.0.

eerhardt committed:
Handle non-SDK projects by running after ResolveNuGetPackageAssets

eiriktsarpalis commented on src/libraries/System.Text.Json/gen/JsonSourceGenerator.3.9.cs:
```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Nit: filename seems a bit odd to me. How about something like `JsonSouceGenerator.Roslyn39.cs`?

eiriktsarpalis replied:
or perhaps `RoslynV3` and `RoslynV4`.

eiriktsarpalis commented on src/libraries/System.Text.Json/gen/Reflection/TypeWrapper.cs:
```diff
@@ -311,9 +311,12 @@ public override FieldInfo[] GetFields(BindingFlags bindingAttr)
+#if ROSLYN4_0_OR_GREATER
                         // symbol represents an explicitly named tuple element
-                        fieldSymbol.IsExplicitlyNamedTupleElement)
+                        || fieldSymbol.IsExplicitlyNamedTupleElement
```
I wish could avoid this. Effectively it will break VS19 sourcegen for any type containing value tuples. Would it be possible to consider potential workarounds for 3.9 or upgrade to 3.11?

One potential workaround would be to check if `fieldSymbol.Name` does not match the `Item[1-7]`.

eiriktsarpalis replied:
This might also provide hints on how the property could be derived:

https://github.com/dotnet/roslyn/blob/315c2e149ba7889b0937d872274c33fcbfe9af5f/src/Compilers/CSharp/Portable/Symbols/Tuples/TupleErrorFieldSymbol.cs#L94

eerhardt replied:
@jcouv @chsienki - any thoughts here? Is there a way we can simulate the `fieldSymbol.IsExplicitlyNamedTupleElement` logic while targeting Roslyn 3.9? Or should we just target 3.10 which has this API?

eerhardt replied:
In an offline discussion with @jaredpar, I decided to upgrade to 3.11. If anyone has concerns/feedback, we can revisit this decision.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System.Text.Json.csproj:
```diff
@@ -349,6 +349,7 @@ System.Text.Json.Utf8JsonReader</PackageDescription>
   </ItemGroup>
   <ItemGroup>
-    <AnalyzerReference Include="..\gen\System.Text.Json.SourceGeneration.csproj" />
+    <AnalyzerReference Include="..\gen\System.Text.Json.SourceGeneration.3.9.csproj" />
```
Similarly, consider using a `.Roslyn3.csproj` naming convention here.

layomia replied:
+1. Or `RoslynV3` as suggested elsewhere so it's clearer what the numbers mean.

eerhardt replied:
I chose to go with:
* System.Text.Json.SourceGeneration.Roslyn3.11.csproj
* System.Text.Json.SourceGeneration.Roslyn4.0.csproj
* System.Text.Json.SourceGeneration.Roslyn3.11.Unit.Tests.csproj
* etc

eerhardt committed:
Respond to PR feedback

- Name .cs and .csproj files with Roslyn in the name
- Upgrade to Roslyn 3.11 so IsExplicitlyNamedTupleElement API is available
- Fix some references to the test projects
- Fix incremental pack of the analyzer targets

layomia reviewed (approved):
LGTM assuming clean CI & there are no issues relating to https://github.com/dotnet/sdk/pull/20793.

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into MultiTargetSGs

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into MultiTargetSGs

eerhardt committed:
Respond to latest changes

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/Microsoft.Extensions.Logging.Abstractions.sln:
```diff
@@ -19,6 +17,12 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{548DF5F7-790
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "ref", "ref", "{7631380A-FB73-4241-9987-0891A21E9769}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Extensions.Logging.Generators.Roslyn4.0", "gen\Microsoft.Extensions.Logging.Generators.Roslyn4.0.csproj", "{A5439E79-96D6-4F02-8DD0-23DFF979851D}"
```
so with future upgrades we have to rename these files?

maryamariyan replied:
if yes, do we prefer to use an intermediary name that doesnt have to change with version future changes?

eerhardt replied:
My thinking is that future upgrades would mean a new file with the appropriate version in the name. If we were no longer targeting a specific version, we would then delete the file for that version. So if were no longer targeting Roslyn v4.0, but instead wanted to target Roslyn v4.1, we would delete the 4.0 file and add a 4.1 file.

Do you have an idea on what an intermediary name pattern would look like that would make sense, but not require the version number in it?

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -437,12 +437,20 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
                                             while (potentialNamespaceParent != null &&
-                                                   potentialNamespaceParent is not NamespaceDeclarationSyntax &&
-                                                   potentialNamespaceParent is not FileScopedNamespaceDeclarationSyntax)
+                                                   potentialNamespaceParent is not NamespaceDeclarationSyntax
```
I think the latest version is using `BaseNamespaceDeclarationSyntax`

see https://github.com/dotnet/runtime/pull/57894/files

eerhardt replied:
These changes are sync'd to that change. Note that in the #57894 change, line 447 uses `BaseNamespaceDeclarationSyntax`, but lines 442-443 use `NamespaceDeclarationSyntax` and `FileScopedNamespaceDeclarationSyntax`, as this change does.

eerhardt commented:
Windows failures are #58927.

Linux_musl x64 failures are infrastructure related.

This has passed CI previously before resolving merge conflicts. Libraries legs are building and tests are passing. Merging to unblock 6.0 work.
